src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/AddUsingsCodeActionResolver.cs

```diff
@@ -0,0 +1,239 @@
+            _documentResolver = documentResolver ?? throw new ArgumentNullException(nameof(documentResolver));
+        }
+
+        public override string Action => LanguageServerConstants.CodeActions.AddUsings;
+
+        public override async Task<WorkspaceEdit> ResolveAsync(JObject data, CancellationToken cancellationToken)
+        {
+            var actionParams = data.ToObject<AddUsingsCodeActionParams>();
+            var path = actionParams.Uri.GetAbsoluteOrUNCPath();
+
+            var document = await Task.Factory.StartNew(() =>
+            {
+                _documentResolver.TryResolveDocument(path, out var documentSnapshot);
+                return documentSnapshot;
+            }, cancellationToken, TaskCreationOptions.None, _foregroundDispatcher.ForegroundScheduler).ConfigureAwait(false);
+            if (document is null)
+            {
+                return null;
+            }
+
+            var text = await document.GetTextAsync().ConfigureAwait(false);
+            if (text is null)
+            {
+                return null;
+            }
+
+            var codeDocument = await document.GetGeneratedOutputAsync().ConfigureAwait(false);
+            if (codeDocument.IsUnsupported())
+            {
+                return null;
+            }
+
+            if (!FileKinds.IsComponent(codeDocument.GetFileKind()))
+            {
+                return null;
+            }
+
+            var codeDocumentIdentifier = new VersionedTextDocumentIdentifier() { Uri = actionParams.Uri, Version = 0 };
+            var documentChanges = new List<WorkspaceEditDocumentChange>();
+
+            var namespaceList = actionParams.Namespaces;
+
+            var usingDirectives = FindUsingDirectives(codeDocument);
+            if (usingDirectives.Count > 0)
+            {
+                documentChanges.Add(GenerateUsingEditsInterpolated(codeDocument, codeDocumentIdentifier, actionParams.Namespaces, usingDirectives));
+            }
+            else
+            {
+                documentChanges.Add(GenerateUsingEditsAtTop(codeDocument, codeDocumentIdentifier, actionParams.Namespaces));
+            }
+
+            return new WorkspaceEdit()
+            {
+                DocumentChanges = documentChanges
+            };
+        }
+
+        private static WorkspaceEditDocumentChange GenerateUsingEditsInterpolated(
+            RazorCodeDocument codeDocument,
+            VersionedTextDocumentIdentifier codeDocumentIdentifier,
+            string[] namespaceArray,
+            List<RazorUsingDirective> usingDirectives)
+        {
+            // Sort and queue namespaces
+            var namespaceList = namespaceArray.ToList();
+            namespaceList.Sort();
+            var namespaces = new Queue<string>(namespaceList);
+
+            var edits = new List<TextEdit>();
+            foreach (var usingDirective in usingDirectives)
```

Comment:
Are we doing interpolation to avoid the performance hit of simply overwriting the existing usings with a new sorted list of usings?

Reply:
Mainly I just want to avoid editing things that are out of our scope. I just assumed this was the right way, I could be wrong!